pkg/webhook/policy_benchmark_test.go

```diff
@@ -0,0 +1,332 @@
+			RequestKind:        &metav1.GroupVersionKind{Group: oldRes.GroupVersionKind().Group, Version: oldRes.GroupVersionKind().Version, Kind: oldRes.GroupVersionKind().Kind},
+			RequestResource:    &metav1.GroupVersionResource{Group: gvr.Group, Version: gvr.Version, Resource: gvr.Resource},
+			RequestSubResource: "",
+			Name:               name,
+			Namespace:          namespace,
+			Operation:          "UPDATE",
+			UserInfo: authenticationv1.UserInfo{
+				Username: "res-creator",
+				UID:      "uid",
+				Groups:   []string{"res-creator-group"},
+				Extra:    map[string]authenticationv1.ExtraValue{"extraKey": {"value1", "value2"}}},
+			Object:    runtime.RawExtension{Object: &resList[n%len(resList)]},
+			OldObject: runtime.RawExtension{Object: oldRes},
+			DryRun:    &dryRun,
+			Options:   runtime.RawExtension{},
+		},
+	}
+}
+
+func BenchmarkValidationHandler(b *testing.B) {
+	// setup test
+	opa, err := makeOpaClient()
+	if err != nil {
+		b.Fatalf("could not initialize OPA: %s", err)
+	}
+
+	c := &fakeNsGetter{scheme: scheme.Scheme, NoopClient: testclient.NoopClient{}}
+	cfg := &v1alpha1.Config{
+		Spec: v1alpha1.ConfigSpec{
+			Validation: v1alpha1.Validation{
+				Traces: []v1alpha1.Trace{},
+			},
+		},
+	}
+	h := validationHandler{opa: opa, client: c, injectedConfig: cfg}
+
+	// M constraints among T templates
+	// M is divided into T, 2T, 10T, 20T...100000T
+	benchmarks := map[string]struct {
+		// description of the test
+		description string
+		// directory to load constraint templates from
+		templateDir string
+		// directory to load constraints from
+		constraintDir string
+		// directory to load resources to be evaluated // ex. Pods
+		resourceDir string
+		// number of constraints to load
+		load []int
+	}{
+		"psp": {
+			description:   "All constraints are applicable and all requests are violating",
+			templateDir:   "testdata/psp-templates",
+			constraintDir: "testdata/psp-constraints",
+			// pod list
+			resourceDir: "testdata/psp-pods",
+			load:        []int{5, 10, 50, 100, 200, 1000, 2000},
+		},
+		// create template, constraint and resource directories and add a new test with appropriate load values
+	}
+	for name, tc := range benchmarks {
+		// read template
+		ctList, err := readTemplates(tc.templateDir)
+		if err != nil {
+			b.Fatalf("failed to read template files: %s", err)
+		}
+
+		// read constraints
+		crList, err := readConstraints(tc.constraintDir)
+		if err != nil {
+			b.Fatalf("failed to read constraint files: %s", err)
+		}
+
+		// read resources
+		resList, err := readResources(tc.resourceDir)
+		if err != nil {
+			b.Fatalf("failed to read resources: %s", err)
+		}
+
+		for _, bm := range tc.load {
+			// remove all data from OPA
+			err = opa.Reset(context.TODO())
+			if err != nil {
+				b.Errorf("test %s, failed to reset OPA: %s", name, err)
+			}
+			// seed random generator
+			rand.Seed(time.Now().UnixNano())
+
+			// create T templates
+			err = addTemplates(opa, ctList)
```

Comment:
Would it be hard to add a benchmark for template/constraint loading here? Looks like we could just call separate b.Run()s?

Reply:
It might be better to add template and constraint loading as a separate benchmark test. It would enable us to use different data set for each , and would be easier to read and extend.
Let me know your thoughts on this.

I can add another benchmark test for these two operations (loading of constraints and templates), which is not too hard.

Reply:
If you think it's best done as a different test and doing so doesn't majorly change the design of what's already here, we can add it later.

Reply:
The overall process will be very similar. We would need to read M templates or generate them, and the operation under test `opa.AddTemplate` would be put under a b.Run() block. Go benchmark will automatically run it for multiple iterations and generate a report of the average time per operation (along with memory usage if so desired). M can be varied from 1 to 1000 (any number desired). Same for constraints.

This resource was helpful to me: https://golang.org/pkg/testing/#hdr-Benchmarks 

Adding another test should not impact the design of this one.